File: runtime/windows/src/control/dispatcher.ts
import { runtimeErrorCode, type RuntimeErrorCode } from "../errors.js";
import type { RecentRequest, RuntimeStateStore } from "../state.js";
import { RuntimeController } from "../controller/runtime-controller.js";
import {
  CONTROL_PROTOCOL_VERSION,
  ControlResponseSchema,
  type ControlDispatchResult,
  type ControlResponse,
} from "./result.js";
import type { ControlRequest } from "./protocol.js";

interface CachedResponse {
  readonly command: ControlRequest["command"];
  readonly response: ControlResponse;
}

interface InFlightRequest {
  readonly command: ControlRequest["command"];
  readonly result: Promise<ControlDispatchResult>;
}

type MutationRequest = Exclude<ControlRequest, { readonly command: "status" }>;

function errorMessage(code: RuntimeErrorCode): string {
  if (code === "RUNTIME_BUSY") return "Another Runtime command is in progress.";
  if (code === "RUNTIME_CONTROL_UNAVAILABLE") return "Runtime control is unavailable.";
  if (code === "RESTORE_AWAITING_EXIT") return "Codex must exit normally before changing themes.";
  if (code === "CODEX_WINDOW_ACTIVATION_FAILED") {
    return "Codex window activation could not be verified safely.";
  }
  if (code === "PROCESS_INSPECTION_DENIED") {
    return "Codex process identity could not be inspected safely.";
  }
  if (code === "ADAPTER_INCOMPATIBLE") {
    return "The installed Codex UI is not compatible with this OpenChatGPTSkin adapter.";
  }
  if (code === "INTERNAL") return "The Runtime command could not be completed.";
  return "The Runtime command was rejected safely.";
}

function errorNextAction(code: RuntimeErrorCode): string {
  if (code === "CODEX_WINDOW_ACTIVATION_FAILED") {
    return "Quit Codex completely and retry once. If it repeats, report " +
      "CODEX_WINDOW_ACTIVATION_FAILED with the installed Codex version.";
  }
  if (code === "PROCESS_INSPECTION_DENIED") {
    return "Quit Codex completely and retry once. If it repeats, report " +
      "PROCESS_INSPECTION_DENIED with the installed Codex version.";
  }
  if (code === "ADAPTER_INCOMPATIBLE") {
    return "Update OpenChatGPTSkin before retrying. If no update is available, report " +
      "ADAPTER_INCOMPATIBLE with the installed Codex version.";
  }
  return "Review Runtime status and retry with a new request ID.";
}

function errorResponse(
  request: ControlRequest,
  code: RuntimeErrorCode,
): ControlResponse {
  return ControlResponseSchema.parse({
    protocolVersion: CONTROL_PROTOCOL_VERSION,
    requestId: request.requestId,
    ok: false,
    error: {
      code,
      message: errorMessage(code),
      nextAction: errorNextAction(code),
    },
  });
}

function successResponse(
  request: ControlRequest,
  result: Awaited<ReturnType<RuntimeController["status"]>>,
): ControlResponse {
  return ControlResponseSchema.parse({
    protocolVersion: CONTROL_PROTOCOL_VERSION,
    requestId: request.requestId,
    ok: true,
    result,
  });
}

export class RuntimeControlDispatcher {
  private mutation: Promise<void> | null = null;
  private readonly cache = new Map<string, CachedResponse>();
  private readonly inFlight = new Map<string, InFlightRequest>();

  constructor(
    private readonly controller: RuntimeController,
    private readonly state: RuntimeStateStore,
  ) {}

  async dispatch(request: ControlRequest): Promise<ControlDispatchResult> {
    try {
      const cached = await this.lookup(request);
      if (cached) return { response: cached };

      const inFlight = this.inFlight.get(request.requestId);
      if (inFlight) {
        if (inFlight.command !== request.command) {
          return this.complete(request, errorResponse(request, "RUNTIME_CONTROL_UNAVAILABLE"));
        }
        return inFlight.result.then(({ response }) => ({ response }));
      }

      if (request.command === "status") {
        return this.executeStatus(request);
      }

      if (this.mutation) {
        return this.complete(request, errorResponse(request, "RUNTIME_BUSY"));
      }

      const result = this.executeMutation(request);
      this.inFlight.set(request.requestId, { command: request.command, result });
      try {
        return await result;
      } finally {
        this.inFlight.delete(request.requestId);
      }
    } catch (error) {
      return this.complete(request, errorResponse(request, runtimeErrorCode(error)));
    }
  }

  private async lookup(request: ControlRequest): Promise<ControlResponse | null> {
    const persistent = await this.state.read();
    const stored = persistent?.recentRequests.find((entry) => entry.requestId === request.requestId);
    if (stored) return this.matchStoredRequest(request, stored);

    const cached = this.cache.get(request.requestId);
    if (!cached) return null;
    if (cached.command !== request.command) {
      return errorResponse(request, "RUNTIME_CONTROL_UNAVAILABLE");
    }
    return cached.response;
  }

  private matchStoredRequest(
    request: ControlRequest,
    stored: RecentRequest,
  ): ControlResponse {
    if (stored.command !== request.command || stored.response.requestId !== request.requestId) {
      return errorResponse(request, "RUNTIME_CONTROL_UNAVAILABLE");
    }
    this.remember(stored.requestId, { command: stored.command, response: stored.response });
    return stored.response;
  }

  private async executeStatus(request: ControlRequest): Promise<ControlDispatchResult> {
    try {
      return await this.complete(request, successResponse(request, await this.controller.status()));
    } catch (error) {
      return this.complete(request, errorResponse(request, runtimeErrorCode(error)));
    }
  }

  private async executeMutation(request: MutationRequest): Promise<ControlDispatchResult> {
    let release!: () => void;
    this.mutation = new Promise<void>((resolve) => { release = resolve; });
    try {
      const result = await this.executeControllerCommand(request);
      const response = successResponse(request, result);
      const afterResponse = request.command === "restore"
        ? await this.terminalExitMonitorCallback()
        : undefined;
      return this.complete(request, response, afterResponse);
    } catch (error) {
      const afterResponse = request.command === "launch"
        ? await this.terminalExitMonitorCallback()
        : undefined;
      return this.complete(
        request,
        errorResponse(request, runtimeErrorCode(error)),
        afterResponse,
      );
    } finally {
      release();
      this.mutation = null;
    }
  }

  private async executeControllerCommand(
    request: MutationRequest,
  ): Promise<Awaited<ReturnType<RuntimeController["status"]>>> {
    switch (request.command) {
      case "launch":
        return this.controller.launch(
          request.params.themeId,
          request.requestId,
          request.params.themeVersion,
        );
      case "switch":
        return this.controller.switchTheme(
          request.params.themeId,
          request.requestId,
          request.params.themeVersion,
        );
      case "pause":
        return this.controller.pause(request.requestId);
      case "resume":
        return this.controller.resume(request.requestId);
      case "restore":
        return this.controller.restore(request.requestId);
    }
  }

  private async terminalExitMonitorCallback(): Promise<ControlDispatchResult["afterResponse"]> {
    const session = await this.state.read();
    if (session?.status !== "restored-awaiting-exit" &&
      session?.status !== "restored-cleanup-required") {
      return undefined;
    }
    return () => this.controller.startExitMonitoring();
  }

  private async complete(
    request: ControlRequest,
    response: ControlResponse,
    afterResponse?: () => Promise<void> | void,
  ): Promise<ControlDispatchResult> {
    const parsed = ControlResponseSchema.parse(response);
    const record: RecentRequest = {
      requestId: request.requestId,
      command: request.command,
      response: parsed,
      completedAt: new Date().toISOString(),
    };
    try {
      await this.state.appendRecentRequest(record);
    } catch (error) {
      const persistenceFailure = errorResponse(request, runtimeErrorCode(error));
      this.remember(request.requestId, {
        command: request.command,
        response: persistenceFailure,
      });
      return { response: persistenceFailure };
    }
    this.remember(request.requestId, { command: request.command, response: parsed });
    return afterResponse ? { response: parsed, afterResponse } : { response: parsed };
  }

  private remember(requestId: string, value: CachedResponse): void {
    this.cache.delete(requestId);
    this.cache.set(requestId, value);
    while (this.cache.size > 32) {
      const oldest = this.cache.keys().next().value;
      if (!oldest) return;
      this.cache.delete(oldest);
    }
  }
}

File: host/go/internal/control/dispatcher.go
package control

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"regexp"
	"sync"
)

type Handler func(context.Context, Request) (Result, error)

type cachedResponse struct {
	command  string
	response Response
}

type inFlight struct {
	command string
	done    chan struct{}
}

type Dispatcher struct {
	handler  Handler
	mu       sync.Mutex
	mutation sync.Mutex
	cache    map[string]cachedResponse
	order    []string
	inFlight map[string]*inFlight
}

func NewDispatcher(handler Handler) *Dispatcher {
	return &Dispatcher{
		handler:  handler,
		cache:    make(map[string]cachedResponse),
		inFlight: make(map[string]*inFlight),
	}
}

var (
	requestIDPattern    = regexp.MustCompile(`^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[1-5][0-9a-fA-F]{3}-[89abAB][0-9a-fA-F]{3}-[0-9a-fA-F]{12}$`)
	themeIDPattern      = regexp.MustCompile(`^[a-z0-9]+(?:-[a-z0-9]+)*$`)
	themeVersionPattern = regexp.MustCompile(`^(?:0|[1-9]\d*)\.(?:0|[1-9]\d*)\.(?:0|[1-9]\d*)$`)
)

type ThemeParameters struct {
	ThemeID      string `json:"themeId"`
	ThemeVersion string `json:"themeVersion,omitempty"`
}

func decodeStrictParameters(params json.RawMessage, target any) error {
	decoder := json.NewDecoder(bytes.NewReader(params))
	decoder.DisallowUnknownFields()
	return decoder.Decode(target)
}

func DecodeThemeParameters(request Request) (ThemeParameters, error) {
	var params ThemeParameters
	if err := decodeStrictParameters(request.Params, &params); err != nil {
		return ThemeParameters{}, errors.New("control theme parameters are invalid")
	}
	if len(params.ThemeID) < 3 || len(params.ThemeID) > 80 || !themeIDPattern.MatchString(params.ThemeID) {
		return ThemeParameters{}, errors.New("control theme ID is invalid")
	}
	if params.ThemeVersion != "" && !themeVersionPattern.MatchString(params.ThemeVersion) {
		return ThemeParameters{}, errors.New("control theme version is invalid")
	}
	return params, nil
}

func validateParameters(request Request) error {
	if request.Command == "launch" || request.Command == "switch" {
		_, err := DecodeThemeParameters(request)
		return err
	}
	var params map[string]json.RawMessage
	if err := decodeStrictParameters(request.Params, &params); err != nil || params == nil || len(params) != 0 {
		return errors.New("control parameters must be an empty object")
	}
	return nil
}

func validateRequest(request Request) error {
	if request.ProtocolVersion != ProtocolVersion {
		return errors.New("control protocol version is invalid")
	}
	if !requestIDPattern.MatchString(request.RequestID) {
		return errors.New("control request ID is invalid")
	}
	switch request.Command {
	case "launch", "status", "switch", "pause", "resume", "restore":
	default:
		return errors.New("control command is invalid")
	}
	if len(request.Params) == 0 || !json.Valid(request.Params) {
		return errors.New("control parameters are invalid")
	}
	return validateParameters(request)
}

func rejected(request Request, code, message string) Response {
	return Response{
		ProtocolVersion: ProtocolVersion,
		RequestID:       request.RequestID,
		OK:              false,
		Error:           &Error{Code: code, Message: message, NextAction: "Review Runtime status and retry with a new request ID."},
	}
}

func (dispatcher *Dispatcher) Dispatch(ctx context.Context, request Request) Response {
	if err := validateRequest(request); err != nil {
		return rejected(request, "RUNTIME_CONTROL_UNAVAILABLE", err.Error())
	}
	for {
		dispatcher.mu.Lock()
		if cached, ok := dispatcher.cache[request.RequestID]; ok {
			dispatcher.mu.Unlock()
			if cached.command != request.Command {
				return rejected(request, "RUNTIME_CONTROL_UNAVAILABLE", "request ID belongs to another command")
			}
			return cached.response
		}
		if pending, ok := dispatcher.inFlight[request.RequestID]; ok {
			dispatcher.mu.Unlock()
			if pending.command != request.Command {
				return rejected(request, "RUNTIME_CONTROL_UNAVAILABLE", "request ID belongs to another command")
			}
			select {
			case <-pending.done:
				continue
			case <-ctx.Done():
				return rejected(request, "RUNTIME_CONTROL_UNAVAILABLE", "request was cancelled")
			}
		}
		pending := &inFlight{command: request.Command, done: make(chan struct{})}
		dispatcher.inFlight[request.RequestID] = pending
		dispatcher.mu.Unlock()
		response := dispatcher.execute(ctx, request)
		dispatcher.mu.Lock()
		dispatcher.cache[request.RequestID] = cachedResponse{command: request.Command, response: response}
		dispatcher.order = append(dispatcher.order, request.RequestID)
		for len(dispatcher.order) > 32 {
			oldest := dispatcher.order[0]
			dispatcher.order = dispatcher.order[1:]
			delete(dispatcher.cache, oldest)
		}
		delete(dispatcher.inFlight, request.RequestID)
		close(pending.done)
		dispatcher.mu.Unlock()
		return response
	}
}

func (dispatcher *Dispatcher) execute(ctx context.Context, request Request) Response {
	if request.Command != "status" {
		dispatcher.mutation.Lock()
		defer dispatcher.mutation.Unlock()
	}
	result, err := dispatcher.handler(ctx, request)
	if err != nil {
		return rejected(request, "INTERNAL", "The Runtime command could not be completed.")
	}
	payload, err := json.Marshal(result)
	if err != nil {
		return rejected(request, "INTERNAL", "The Runtime result could not be encoded.")
	}
	return Response{ProtocolVersion: ProtocolVersion, RequestID: request.RequestID, OK: true, Result: payload}
}
